File: web/src/lib/external.ts
// Opening a link outside the cockpit.
//
// The cockpit runs in two hosts — a plain browser tab and the native webview
// shell (ADR 0013) — and "open this URL" means something different in each. A
// browser tab can open another tab; the shell's window is the whole application,
// so opening a link *inside* it would navigate away from the cockpit with no way
// back. The shell therefore hands the page a global that reaches the operator's
// real browser, and this module is the one place that decides between the two.
//
// The contract is the `__chartrTitleBar` one (see titlebar.ts): the shell injects
// a global before the page loads, and its mere presence is the capability. A
// plain browser tab never sees it and takes the fallback, so the fallback — not
// the hook — is the behaviour everywhere the shell is not, including before the
// macOS shell addition ships.
declare global {
  interface Window {
    // Bound by the native shell. Takes the URL and hands it to the OS; the
    // shell's binding returns a promise, which we neither await nor need.
    __chartrOpenExternal?: (url: string) => unknown
  }
}

// Where a link went, returned so the decision is observable from a test without
// a real window or a real shell.
//
//   - `shell`   — handed to the native shell's hook, which opens the system browser.
//   - `browser` — opened in a new browser tab, the no-shell fallback.
//   - `blocked` — not a scheme we are willing to hand to either.
export type OpenOutcome = 'shell' | 'browser' | 'blocked'

// Only ever hand out http(s). The web-links addon only matches those anyway, so
// this guard is not what makes links work — it is what keeps this function from
// becoming a way for terminal *output* to launch something on the operator's
// machine. `open`/`xdg-open` on the shell side will happily act on a `file:` or
// an app's custom scheme, and terminal output is untrusted text from whatever
// the agent printed, so the narrow allowlist lives on both sides of the seam.
function isOpenable(url: string): boolean {
  try {
    const scheme = new URL(url).protocol
    return scheme === 'http:' || scheme === 'https:'
  } catch {
    return false
  }
}

/**
 * Open a URL outside the cockpit, preferring the native shell's hook and falling
 * back to a new browser tab. Returns which path it took.
 *
 * `win` is injectable so the decision is a pure unit: pass a stub carrying (or
 * omitting) `__chartrOpenExternal` and assert the outcome.
 */
export function openExternal(url: string, win: Window = window): OpenOutcome {
  if (!isOpenable(url)) return 'blocked'

  const hook = win.__chartrOpenExternal
  if (typeof hook === 'function') {
    // The shell's binding answers with a promise that rejects when the OS opener
    // refused. Nothing here can retry and the operator has no action to take, so
    // it is swallowed rather than left as an unhandled rejection in the console.
    void Promise.resolve(hook(url)).catch(() => {})
    return 'shell'
  }

  // `noopener` is the standard hygiene for a link out of the app: the new tab
  // gets no `window.opener` handle back into the cockpit.
  win.open(url, '_blank', 'noopener,noreferrer')
  return 'browser'
}

File: cmd/webview/external.go
package main

import (
	"fmt"
	"net/url"
	"os/exec"
	"runtime"
)

// Opening a link in the operator's real browser.
//
// The shell's window is the whole application, so a URL clicked in terminal
// output must not navigate it — the operator would lose the cockpit with no way
// back. `main_webview.go` binds openExternalURL into the page as
// `window.__chartrOpenExternal`, mirroring the `__chartrTitleBar` contract: the
// shell injects a global, the page treats its presence as the capability, and a
// plain browser tab that never sees it opens a new tab instead (web/src/lib/
// external.ts). This file is untagged so the URL guard compiles and is tested by
// the default cgo-free build, exactly as the lock is (ADR 0013).

// openExternalURL hands a URL to the OS to open in the operator's default
// browser. It never blocks on the child — the browser is the operator's to close.
//
// It is bound to the page, so its argument is whatever the cockpit passed, which
// ultimately traces back to text an agent printed into a terminal. Untrusted
// input reaching `open` is the risk worth naming: `open` and `xdg-open` will
// launch an application for a `file:` path or a registered custom scheme, so the
// guard below is not a formality.
func openExternalURL(raw string) error {
	if err := checkExternalURL(raw); err != nil {
		return err
	}
	opener := osOpener()
	if opener == "" {
		return fmt.Errorf("no way to open a URL on %s", runtime.GOOS)
	}
	cmd := exec.Command(opener, raw)
	if err := cmd.Start(); err != nil {
		return err
	}
	// Reap it in the background so a browser launched from here is not left a
	// zombie child of the shell.
	go func() { _ = cmd.Wait() }()
	return nil
}

// checkExternalURL accepts only an absolute http(s) URL. Everything else — a
// `file:` path, a custom app scheme, a bare string, a scheme-relative reference —
// is refused rather than passed to the OS opener. The client keeps the same
// allowlist; both sides hold it because either alone would be the only thing
// standing between terminal output and a launched application.
func checkExternalURL(raw string) error {
	u, err := url.Parse(raw)
	if err != nil {
		return fmt.Errorf("not a URL: %w", err)
	}
	if u.Scheme != "http" && u.Scheme != "https" {
		return fmt.Errorf("refusing to open %q: only http and https URLs are opened", raw)
	}
	if u.Host == "" {
		return fmt.Errorf("refusing to open %q: no host", raw)
	}
	return nil
}

// osOpener is the platform's "open this in whatever handles it" command — the
// same ladder the config surface uses to open a file in an editor
// (internal/server/configsurface.go). Duplicated rather than shared: it is three
// lines, and the shell must not import the server's unexported internals.
func osOpener() string {
	switch runtime.GOOS {
	case "darwin":
		return "open"
	case "windows":
		return "explorer"
	default:
		return "xdg-open"
	}
}
